firstuseauthenticator/firstuseauthenticator.py

```diff
@@ -10,17 +10,24 @@
 from jinja2 import ChoiceLoader, FileSystemLoader
 from jupyterhub.auth import Authenticator
 from jupyterhub.handlers import BaseHandler
+from jupyterhub.handlers import LoginHandler
 from jupyterhub.orm import User
 
 from tornado import gen, web
-from traitlets.traitlets import Unicode, Bool
+from traitlets.traitlets import Unicode, Bool, Integer
 
 import bcrypt
 
 
 TEMPLATE_DIR = os.path.join(os.path.dirname(__file__), 'templates')
 
 
+class CustomLoginHandler(LoginHandler):
+    custom_login_error = 'Invalid username or password'
+    def _render(self, login_error=None, username=None):
+        return super()._render(self.custom_login_error, username)
+
```

Comment:
A brief docstring here on the purpose would be useful as Custom is very generic prefix.

I understand from the code that we ignore any passed login_error argument and replace it with custom_login_error. I'm reviewing this in the dark without good insight here, does it make sense to ignore login_error entirely or should the login_error parameter instead have a new default value of custom_login_error?

Reply:
I took another look at the code and remembered why I chose to ignore `login_error`.
* In order to customize the `login_error` that gets rendered, I had three options (or at least that's how many I managed to identify):
  1. Override the `post` method of [`class LoginHandler`](https://github.com/jupyterhub/jupyterhub/blob/f97ef7eaac73e4af4a5444b24fdffe47f388da70/jupyterhub/handlers/login.py#L137) (the place where the error gets rendered in the UI).
  2. Have `login_error` be a `LoginHandler` class member (in JupyterHub repo) and change it when appropriate (right now it's just a parameter passed to the function `_render`). 
  3. Inherit from `LoginHandler` and have a `custom_login_error` that replaces `login_error` in the manner described at option 2.

* The reasons why I chose option 3:
  1. **Option 1** would have meant a lot of duplicated code because the `post` method would have been pretty much the same, except for [this line](https://github.com/jupyterhub/jupyterhub/blob/f97ef7eaac73e4af4a5444b24fdffe47f388da70/jupyterhub/handlers/login.py#L153).
  2. **Option 2** would have mean messing with the code in JH and I didn't think it was a good idea at that point.
   3. I went with **option 3** because it felt like the most straightforward approach and it didn't need any outside the repo intervention.

Because the default value of `custom_login_error` is set to be [the only value `login_error` can hold](https://github.com/jupyterhub/jupyterhub/blob/f97ef7eaac73e4af4a5444b24fdffe47f388da70/jupyterhub/handlers/login.py#L153) (i.e. *Invalid username or password'* ), `login_error` can be safely ignored without losing any information.
If I would have given `login_error` a new default value of `custom_login_error`, the `post` method call would have overwritten it with *Invalid username or password*. 

What are your thoughts on this @consideRatio?

P.S. A docstring and probably more comments too are definitely needed :sun_with_face:

Reply:
Ah excellent writedown! So good that you perhaps could reference it by a link in the docstring where you keep it brief about the purpose of the class for example.

You seem to have considered this very thoroughly, given that and I think we could merge this when you are done and release 0.14.0. What do you think?